provider/balance_checker.go

```diff
@@ -113,7 +211,58 @@ loop:
 			balanceCheckResult = runner.Do(func() runner.Result {
 				return runner.NewResult(bc.doCheck(ctx))
 			})
+		case evt := <-subscriber.Events():
+			switch ev := evt.(type) {
+			case event.LeaseAddFundsMonitor:
+				bc.leases[ev.LeaseID] = nil
+				if bc.cfg.LeaseFundCheckInterval > 0 {
+					bc.runEscrowCheck(ctx, ev.LeaseID, leaseCheckCh)
+				}
+			case event.LeaseRemoveFundsMonitor:
+				tm, exists := bc.leases[ev.LeaseID]
+				if !exists {
+					break
+				}
+
+				delete(bc.leases, ev.LeaseID)
+
+				if tm != nil && !tm.Stop() {
+					<-tm.C
+				}
+			}
+		case res := <-leaseCheckCh:
+			// we may have timer fired just a heart beat ahead of lease remove event.
+			if _, exists := bc.leases[res.lid]; !exists {
+				break
+			}
+
+			if res.err != nil {
+				bc.log.Info("couldn't check lease balance. retrying in 1m", "leaseId", res.lid, "error", res.err.Error())
+				bc.leases[res.lid] = time.AfterFunc(time.Minute, func() {
+					select {
+					case <-bc.lc.Done():
+					case leaseCheckCh <- bc.doEscrowCheck(ctx, res.lid):
+					}
+				})
+				break
+			}
+
+			if res.checkAfter == 0 {
```

Comment:
can we just add a function receiver to the response type like "res.isOutOfFunds()" so this is easier to identify? It's checking a duration for a value at this time